lightning/src/util/persist.rs

```diff
@@ -64,41 +114,800 @@ impl<'a, A: KVStorePersister, M: Deref, T: Deref, ES: Deref, NS: Deref, SP: Dere
+				.map_err(|e| KVStoreUpdatingPersisterError::StorageReadFailed {
+					reason: e,
+					context: key,
+				})?,
+			(&*entropy_source, &*signer_provider),
+		) {
+			Ok((blockhash, channel_monitor)) => {
+				if channel_monitor.get_funding_txo().0.txid != outpoint.txid
+					|| channel_monitor.get_funding_txo().0.index != outpoint.index
+				{
+					return Err(KVStoreUpdatingPersisterError::MonitorDecodeFailed {
+						reason: DecodeError::InvalidValue,
+						context: key,
+					}
+					.into());
+				}
+				Ok((blockhash, channel_monitor))
+			}
+			Err(e) => Err(KVStoreUpdatingPersisterError::MonitorDecodeFailed {
+				reason: e,
+				context: key,
+			}
+			.into()),
+		}
+	}
+
+
+
+	/// Deserialize a channel monitor update.
+	fn deserialize_monitor_update(
+		&self,
+		monitor_name: &MonitorName,
+		update_name: &UpdateName,
+	) -> io::Result<ChannelMonitorUpdate> {
+		let ns = self.monitor_update_namespace(monitor_name);
+		let key = update_name.storage_key();
+		Ok(
+			ChannelMonitorUpdate::read(&mut self.kv.read(&ns, &key).map_err(|e| {
+				KVStoreUpdatingPersisterError::StorageReadFailed {
+					reason: e,
+					context: key,
+				}
+			})?)
+			.map_err(|e| KVStoreUpdatingPersisterError::UpdateDecodeFailed {
+				reason: e,
+				context: key,
+			})?,
+		)
+	}
+
+	/// Delete updates with an update_id lower than the given channel monitor.
+	fn delete_stale_updates<ChannelSigner: WriteableEcdsaChannelSigner>(
+		&self,
+		channel_id: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+	) -> io::Result<()> {
+		let monitor_name: MonitorName = channel_id.into();
+		let update_names = self.list_update_names(&monitor_name)?;
+		for update_name in update_names {
+			let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
+			if update.update_id != CLOSED_CHANNEL_UPDATE_ID
+				&& update.update_id <= monitor.get_latest_update_id()
+			{
+				let ns = self.monitor_update_namespace(&monitor_name);
+				let key = update_name.storage_key();
+				self.kv.remove(&ns, key)?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl<ChannelSigner: WriteableEcdsaChannelSigner, K: Deref, L: Deref> Persist<ChannelSigner>
+	for KVStoreUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	// TODO: We really need a way for the persister to inform the user that its time to crash/shut
+	// down once these start returning failure.
+	// A PermanentFailure implies we should probably just shut down the node since we're
+	// force-closing channels without even broadcasting!
+	fn persist_new_channel(
+		&self,
+		funding_txo: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+		_update_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		let key = match MonitorName::try_from(funding_txo) {
+			Ok(monitor_name) => monitor_name.0,
+			Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+		};
+		match self.kv.write(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			&key,
+			&monitor.encode(),
```

Comment:
Had to think about this for a minute -- is it that there's risk of having unapplied updates on disk that you'd load the monitors without?

Doing this makes it _hard_ to downgrade/reverse, but you can do it if you're willing to write a little code to trim off those leading `0xff` bytes.

Reply:
Hey @TheBlueMatt, I did this but PTAL.

Adding a sentinel byte sequence was easy. Removing it before parsing was easy. But doing that blindly made migration _into_ this storage struct impossible (since you'd strip the bytes from an existing, known-full monitor). So I wrote something to read the sentinel, and chain either those bytes or empty bytes to the front of the remaining bytes, depending on whether they match the sentinel or not. Was kinda tricky Rust, so I'd like another set of eyes.

Would also be nice to test the migration by having the standard persister write, then have this one read it, somehow.

Reply:
This is now refactored and is no longer clever. Since `KVStore::read` returns a `Vec<u8>` now, an in-memory cursor is seekable.

Reply:
Oh, I did make it more descriptive -- a 28-byte ASCII string instead of two `0xff`s. Let me know if that's not desirable. IME, an additional 26 bytes will be very little additional I/O for the entire monitors.